File: built-in plugins dsp/mb-dist-tape.cpp
/**
 * MB Tape Saturation
 * Category : effect
 * Type     : distortion
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Analog tape warmth
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_DIST_TAPE_H
#define MB_DIST_TAPE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbDistTape : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-dist-tape";
    static constexpr const char* PLUGIN_NAME    = "MB Tape Saturation";
    static constexpr const char* PLUGIN_TYPE    = "distortion";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float saturation = 0.4f;  // range [0, 1]
    float bias = 0.5f;  // range [0, 1]
    float mix = 1f;  // range [0, 1]
    };

    MbDistTape() = default;
    ~MbDistTape() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.saturation = std::clamp(params.saturation, 0f, 1f);
        params.bias = std::clamp(params.bias, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Tape Saturation
        return input;
    }
};

#endif // MB_DIST_TAPE_H

File: built-in plugins dsp/mb-dist-tape.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbDistTapePlugin: PluginDefinition = {
  id: "mb-dist-tape",
  slug: "mb-dist-tape",
  name: "MB Tape Saturation",
  category: "effect",
  type: "distortion",
  version: "1.0.0",
  description: "Analog tape warmth",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "saturation",
      name: "Saturation",
      type: "float",
      defaultValue: 0.4,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "bias",
      name: "Bias",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "mix",
      name: "Mix",
      type: "float",
      defaultValue: 1,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { saturation: 0.4, bias: 0.5, mix: 1 },
};

export default MbDistTapePlugin;
